pallets/pooled-staking/src/tests/rewards.rs

```diff
@@ -0,0 +1,579 @@
+                delegators_manual: 2_666_250,
+                // Rewards towards auto deleg
+                // = Rewards deleg - Rewards manual deleg
+                // = 8_000_000 - 2_666_250
+                // = 5_333_750
+                delegators_auto: 5_333_750,
+            },
+        );
+    });
+}
+
+#[test]
+fn candidate_only_no_stake() {
+    // Rewarding a candidate that does not have any stake works
+    ExtBuilder::default().build().execute_with(|| {
+        test_distribution(
+            &[],
+            RewardRequest {
+                collator: ACCOUNT_CANDIDATE_1,
+                rewards: 1_000_000,
+            },
+            &[],
+            Distribution {
+                collator_auto: 0,
+                collator_manual: 1_000_000, // 100% of rewards
+                delegators_auto: 0,
+                delegators_manual: 0, // 0% of rewards
+            },
+        )
+    });
+}
+
+#[test]
+fn delegator_only_candidate_zero() {
+    // Rewarding a candidate that does not have any stake works
+    ExtBuilder::default().build().execute_with(|| {
+        test_distribution(
+            &[Delegation {
+                candidate: ACCOUNT_CANDIDATE_1,
+                delegator: ACCOUNT_DELEGATOR_1,
+                pool: TargetPool::ManualRewards,
+                stake: 250_000_000,
+            }],
+            RewardRequest {
+                collator: ACCOUNT_CANDIDATE_1,
+                rewards: 1_000_000,
+            },
+            &[DelegatorState {
+                candidate: ACCOUNT_CANDIDATE_1,
+                delegator: ACCOUNT_DELEGATOR_1,
+                auto_shares: 0,
+                auto_stake: 0,
+                manual_shares: 250,
+                manual_stake: 250_000_000,
+                pending_rewards: 800_000,
+            }],
+            Distribution {
+                collator_auto: 0,
+                collator_manual: 200_000, // 20% of rewards
+                delegators_auto: 0,
+                delegators_manual: 800_000, // 80% of rewards
+            },
+        )
+    });
+}
+
+#[test]
+fn delegator_only_candidate_no_stake_auto_compounding() {
+    // Rewarding a candidate that does not have any stake, while some delegator
+    // has stake for that candidate
+    ExtBuilder::default().build().execute_with(|| {
+        test_distribution(
+            &[Delegation {
+                candidate: ACCOUNT_CANDIDATE_1,
+                delegator: ACCOUNT_DELEGATOR_1,
+                pool: TargetPool::AutoCompounding,
+                stake: 250_000_000,
+            }],
+            RewardRequest {
+                collator: ACCOUNT_CANDIDATE_1,
+                rewards: 1_000_000,
+            },
+            &[],
+            Distribution {
+                collator_auto: 0,
+                collator_manual: 200_000, // 20% of rewards
+                delegators_auto: 800_000, // 80% of rewards
+                delegators_manual: 0,
+            },
```

Comment:
This case is a bit dangerous if we take invulnerables into account. Since invulnerables will have 0 stake, if we use pallet_staking to reward them it will only work as expected only if they have no delegators. If some attacked decides to delegate the min amount to all the invulnerables, this attacker will be able to get 80% of the block rewards. I assume we don't want invulnerables to have delegators?

Reply:
The reward distribution piloting pallet should not call this pallet distribute_rewards for an invulnerable.

Reply:
Ok, and what if a regular collator calls `request_undelegate`? They will still be able to produce blocks until the next session, and they will only have "leaving" shares, which are not taken into account in this algorithm, so it is the same case: a collator with 0 shares that may have delegators.

Reply:
If the candidate wants to leave it seems fair they don't get rewards (outside of the dedicated 20%). Thanks to the joining period one cannot notice a candidate leaving and join to snap the rewards.

Reply:
Yes, but it feels a bit strange. Because there is a small incentive for a delegator to delegate the min amount for all the candidates, because when a candidate leaves they can get 80% of the rewards, if there are no other delegators.

Reply:
the rest of the delegators keep getting the delegation rewards until the cnadidate is kicked out (in 1+session) right?